src/shapes/sdfgrid.cpp

```diff
@@ -494,57 +601,349 @@ class SDFGrid final : public SDF<Float, Spectrum> {
+                 d_z * (k4 + dr::fmadd(k5, o_x, dr::fmadd(k6, o_y, k7 * m0)));
+            c2 = dr::fmadd(
+                m1, m5,
+                d_z * (dr::fmadd(k5, d_x, dr::fmadd(k6, d_y, k7 * m2))));
+            c3 = k7 * m1 * d_z;
+        }
+
+        auto [hit, t] = sdf_solve_cubic(t_beg, t_end, c3, c2, c1, c0);
+
+        if (m_watertight) {
+            FloatP eval_sdf             = -(c3 * t_beg * t_beg * t_beg +
+                                c2 * t_beg * t_beg + c1 * t_beg + c0);
+            dr::masked(t, eval_sdf < 0) = t_beg;
+            hit                         = hit || eval_sdf < 0;
+        }
+
+        active = active && bbox_hit && hit && t >= 0.f && t <= ray.maxt;
+
+        return { active, dr::select(active, t, dr::Infinity<FloatP>),
+                 Point<FloatP, 2>(0, 0), ((uint32_t) -1), prim_index };
+    }
+
+    /* \brief Solve cubic polynomial that gives solution to voxel intersection
+     *
+     *  M ARMITT, G., K LEER , A., WALD , I., AND F RIEDRICH , H.
+     *  2004. Fast and accurate ray-voxel intersection techniques for
+     *  iso-surface ray tracing.
+     */
+    template <typename FloatP>
+    MI_INLINE std::tuple<dr::mask_t<FloatP>, FloatP>
+    sdf_solve_cubic(FloatP t_beg, FloatP t_end, FloatP c3, FloatP c2, FloatP c1,
+                    FloatP c0) const {
+
+        using MaskP = dr::mask_t<FloatP>;
+
+        auto [has_derivative_roots, root_0, root_1] =
+            math::solve_quadratic(c3 * 3, c2 * 2, c1);
+
+        auto eval_sdf_t = [&](FloatP t_) -> FloatP {
+            return -(c3 * t_ * t_ * t_ + c2 * t_ * t_ + c1 * t_ + c0);
+        };
+
+        auto numerical_solve = [&](FloatP t_near, FloatP t_far, FloatP f_near,
+                                   FloatP f_far) -> FloatP {
+            static constexpr uint32_t num_solve_max_iter = 50;
+            static constexpr float num_solve_epsilon     = 0.004f;
+
+            FloatP t   = 0;
+            FloatP f_t = 0;
+
+            uint32_t i = 0;
+            MaskP done = false;
+            while (!dr::all(done)) {
+                t   = t_near + (t_far - t_near) * (-f_near / (f_far - f_near));
+                f_t = eval_sdf_t(t);
+                FloatP condition = f_t * f_near;
+                t_far            = dr::select(condition <= 0, t, t_far);
+                f_far            = dr::select(condition <= 0, f_t, f_far);
+
+                t_near = dr::select(condition > 0, t, t_near);
+                f_near = dr::select(condition > 0, f_t, f_near);
+                done   = (dr::abs(f_t) < num_solve_epsilon) ||
+                       (num_solve_max_iter < ++i);
+            }
+
+            return t;
+        };
+
+        FloatP t_near = t_beg;
+        FloatP t_far  = t_end;
+
+        FloatP f_root_0 = eval_sdf_t(root_0);
+        FloatP f_root_1 = eval_sdf_t(root_1);
+
+        MaskP root_0_valid = t_near <= root_0 && root_0 <= t_far;
+
+        dr::masked(t_far, has_derivative_roots && root_0_valid &&
```

Comment:
`dr::masked` is only needed if the left hand side is potentially *not* a dr.jit array.
You may prefer the slice assignment syntax for Dr.Jit arrays:

`array[mask] = ...;`.

Reply:
Thanks! Makes sense. I'll clean that up.